### PR TITLE
updating extra info logic

### PR DESCRIPTION
**Pull request by chrisnojima**
@MarcoPolo  i'm halfway done. doing the group view next but this does the search results. take a peek tnx

### CODE REVIEW

File: shared/actions/search.js [hunk 1]
```diff
@@ -31,50 +32,47 @@ type RawResult = {
...
-}
-
 function parseExtraInfo (platform: ?SearchPlatforms, rr: RawResult, isFollowing: (username: string) => boolean): ExtraInfo {
-  const fullName = parseFullName(rr)
-  const serviceName = rr.service && rr.service.service_name && capitalize(rr.service.service_name)
+  const serviceName = rr.service && capitalize(rr.service.service_name || '')
 
   if (platform === 'Keybase') {
-    if (serviceName) {
+    if (rr.service) {
```
Comment (MarcoPolo) [inline shared/actions/search.js]: Why the change from `serviceName`?

File: shared/actions/search.js [hunk 2]
```diff
@@ -31,50 +32,47 @@ type RawResult = {
   }
 }
 
-function parseFullName (rr: RawResult): string {
```
Comment (MarcoPolo) [inline shared/actions/search.js]: I like this better, you're duplicating login on how to read full_name, this will give you the full_name in one spot, what's the reasoning to remove this?

File: shared/constants/search.js [hunk 3]
```diff
@@ -7,10 +7,10 @@ export type SearchPlatforms = 'Keybase' | 'Twitter' | 'Github' | 'Reddit' | 'Coi
 
 export type ExtraInfo = {
   service: 'external',
-  icon: IconType,
+  icon: ?IconType,
   serviceUsername: string, // i.e. with twitter it would be malgorithms
-  serviceAvatar: ?string, // i.e. with twitter it would be their twitter avatar url
-  fullNameOnService: ?string // Say with twitter we know malgorithms is "Chris Coyne"
+  serviceAvatar: string, // i.e. with twitter it would be their twitter avatar url
+  fullNameOnService: string // Say with twitter we know malgorithms is "Chris Coyne"
```
Comment (MarcoPolo) [inline shared/constants/search.js]: I think we should add another variant that will be used when we only show the icon, username, and full name. We could call it `externalHint`

This variant would change to only contain avatar, username, and fullname.

File: shared/actions/search.js (hunk 1 above)
Comment (chrisnojima) [inline shared/actions/search.js]: really this code cares about the structure of the incoming rawResult objet and not a piece of derived data. if we have a service sub object we want to do this flow. if capitalize() happened to have something like if (service_name === twitter) return '' then we'd do the wrong thing (for example)

File: shared/search/user-pane/index.js [hunk 4]
```diff
@@ -49,7 +49,7 @@ export default connector.connect(
       return {
         mode: 'external',
         nonUserInfoProps: {
-          avatar: userForInfoPane.serviceAvatar,
+          avatar: userForInfoPane.extraInfo.serviceAvatar || '',
```
Comment (MarcoPolo) [inline shared/search/user-pane/index.js]: why this change?

you aren't sure extraInfo will be the variant that contains the serviceAvatar

File: shared/actions/search.js (hunk 2 above)
Comment (chrisnojima) [inline shared/actions/search.js]: This function is pulling the name from two different data sources so you
can't tell what your going to get. We explicitly want one or the other so
this fallback doesn't match our logic

File: shared/search/user-search/render.native.js [hunk 5]
```diff
@@ -42,13 +45,13 @@ function KeybaseExtraInfo ({username, fullName, isFollowing, searchText}) {
   )
 }
 
-// TODO(MM) use serviceAvatar
-function ExternalExtraInfo ({serviceUsername, fullNameOnService, icon, searchText}) {
+function ExternalExtraInfo ({fullNameOnService, icon, serviceAvatar, serviceUsername, searchText}) {
```
Comment (MarcoPolo) [inline shared/search/user-search/render.native.js]: If we have the externalHint and external variants of extra info, we'd have two simple functions instead of trying to show an icon or avatar.

File: shared/actions/search.js (hunk 2 above)
Comment (MarcoPolo) [inline shared/actions/search.js]: 👍

File: shared/constants/search.js [hunk 6]
```diff
@@ -7,10 +7,10 @@ export type SearchPlatforms = 'Keybase' | 'Twitter' | 'Github' | 'Reddit' | 'Coi
 
 export type ExtraInfo = {
   service: 'external',
-  icon: IconType,
+  icon: ?IconType,
   serviceUsername: string, // i.e. with twitter it would be malgorithms
-  serviceAvatar: ?string, // i.e. with twitter it would be their twitter avatar url
-  fullNameOnService: ?string // Say with twitter we know malgorithms is "Chris Coyne"
+  serviceAvatar: string, // i.e. with twitter it would be their twitter avatar url
```
Comment (MarcoPolo) [inline shared/constants/search.js]: We might want to know when something could be empty. We aren't guaranteed to have a serviceAvatar here, so maybe this should stay `?string`

Comment (chrisnojima) [inline shared/constants/search.js]: we can use '' instead of null then we can have a simpler type right, its always a string and we can still do the same truthy checks against it

File: shared/constants/search.js (hunk 3 above)
Comment (chrisnojima) [inline shared/constants/search.js]: this doesn't happen in practice casue you only get this if you search on an external service and in that case we don't duplicate the username

File: shared/search/user-pane/index.js (hunk 4 above)
Comment (chrisnojima) [inline shared/search/user-pane/index.js]: this pr was only supposed to be for the search results so this was just an unintended leftover

File: shared/search/user-search/render.native.js (hunk 5 above)
Comment (chrisnojima) [inline shared/search/user-search/render.native.js]: i honestly would rather go the other direction, having one component thats driven by props that draws itself correctly instead of 4 variants that are very similar. we already have too much duplication in this file (and also native is also super close and should be hoisted anyways)

File: shared/search/user-pane/index.js (hunk 4 above)
Comment (chrisnojima) [inline shared/search/user-pane/index.js]: so this is actually working. if its external we show the external avatar which is in the extra info

Comment (chrisnojima) [inline shared/search/user-pane/index.js]: i think in general this data structure in the store is a bit confusing as its very view centric and it really changes depending on the context. it might be easier to reason about if we had a flatter structure about the user like

```
export type SearchResult = {
  keybaseUsername: string,
  keybaseFullName: string,
  keybaseIsFollowing: boolean,
  serviceIcon: IconType,
  servieAvatar: string,
  serviceUsername: string,
  serviceFullName: SearchPlatforms,
  serviceProfileUrl: string,
}
```

then your logic for which data goes on which side would be handled by your renderers depending on the search platform. right now the data can switch sides and have different combinations, and the 'upgrade' code path also feels wonky. i'm fine to leave this for now but this seems simpler to me

**Comment by chrisnojima** [comment]
@MarcoPolo this handles the rest of the stuff i wanted to do w/ user groups also. lets chat

**Comment by chrisnojima** [comment]
@MarcoPolo bump

File: shared/constants/search.js (hunk 6 above)
Comment (MarcoPolo) [inline shared/constants/search.js]: `''` is a string, but semantically it means we don't have that data. I'm just leaning to making it explicit that we could not have the data versus implying we won't have that data as an empty string

**Comment by MarcoPolo** [comment]
👍
